### PR TITLE
Static in-out stabilization of the Benders algorithm

### PR DESCRIPTION
**Pull request by tbittar**
Implement the in-out stabilization technique to speed up the convergence of the Benders algorithm

### CODE REVIEW

File: src/cpp/benders/logger/CandidateLog.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ std::string CandidateLog::getMainBodyString(const LogData &data) {
 }
 
 void CandidateLog::set_values_and_sizes(const LogData &data) {
-  for (const auto &pairVarnameValue : data.x0) {
+  for (const auto &pairVarnameValue : data.x_cut) {
```
Comment (a-zakir) [inline src/cpp/benders/logger/CandidateLog.cpp:39]:
```
 for (const auto &[name, value] : data.x_cut) {
   
    value_map valuesMap;
    valuesMap[CANDIDATE] = name;
```

Comment (a-zakir) [inline src/cpp/benders/logger/CandidateLog.cpp]:
```suggestion
  for (const auto &[name, value] : data.x_cut) {

    value_map valuesMap;
    valuesMap[CANDIDATE] = name;
    valuesMap[INVEST] =
        get_formatted_string_from_value(value);
    valuesMap[INVEST_MIN] =
        get_formatted_string_from_value(data.min_invest.at(name));
    valuesMap[INVEST_MAX] =
get_formatted_string_from_value(data.max_invest.at(name));
```

File: src/cpp/benders/benders_sequential/BendersSequential.cpp [hunk 2]
```diff
@@ -116,7 +119,7 @@ void BendersSequential::run() {
 
     set_timer_master(timer_master.elapsed());
     _data.elapsed_time = GetBendersTime();
-    _data.stop = stopping_criterion();
+    _data.stop = should_benders_stop();
```
Comment (a-zakir) [inline src/cpp/benders/benders_sequential/BendersSequential.cpp]:
```suggestion
    _data.stop = ShouldBendersStop();
```

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 3]
```diff
@@ -184,14 +187,15 @@ void BendersMpi::write_exception_message(const std::exception &ex) const {
...
+    compute_ub();
     update_best_ub();
     _logger->log_at_iteration_end(bendersDataToLogData(_data));
 
     update_trace();
 
     _data.elapsed_time = GetBendersTime();
     set_timer_master(timer_master.elapsed());
-    _data.stop = stopping_criterion();
+    _data.stop = should_benders_stop();
```
Comment (a-zakir) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp]:
```suggestion
    _data.stop = ShouldBendersStop();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 4]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
   virtual void run() = 0;
   virtual void initialize_problems() = 0;
   virtual void init_data();
-  void print_csv();
   void update_best_ub();
-  bool stopping_criterion();
+  bool should_benders_stop();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  bool ShouldBendersStop();
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 5]
```diff
@@ -209,8 +184,9 @@ void BendersBase::bound_simplex_iter(int simplexiter) {
 /*!
  *  \brief Check if initial relaxation should stop
  */
-bool BendersBase::relaxation_stopping_criterion() const {
-  return (((_data.best_ub - _data.lb) / _data.best_ub) <= _options.RELAXED_GAP);
+bool BendersBase::should_relaxation_stop() const {
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
bool BendersBase::ShouldRelaxationStop() const {
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 6]
```diff
@@ -219,10 +195,11 @@ bool BendersBase::relaxation_stopping_criterion() const {
  *  Method updating the stopping criterion and reinitializing some datas
  *
  */
-bool BendersBase::stopping_criterion() {
+void BendersBase::update_stopping_criterion() {
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
void BendersBase::updateStoppingCriterion() {
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 7]
```diff
@@ -247,28 +228,29 @@ void BendersBase::update_trace() {
...
 bool BendersBase::is_initial_relaxation_requested() const {
   return (_options.MASTER_FORMULATION == MasterFormulation::INTEGER &&
-          _options.INITIAL_MASTER_RELAXATION);
+          _options.SEPARATION_PARAM < 1);
 }
 
 bool BendersBase::switch_to_integer_master(bool is_relaxed) const {
   return is_initial_relaxation_requested() && is_relaxed &&
-         relaxation_stopping_criterion();
+         should_relaxation_stop();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
         ShouldRelaxationStop();
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 8]
```diff
@@ -352,6 +333,36 @@ void BendersBase::activate_integrity_constraints() const {
   _master->activate_integrity_constraints();
 }
 
+void BendersBase::compute_x_cut() {
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
void BendersBase::ComputeXCut() {
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 9]
```diff
@@ -352,6 +333,36 @@ void BendersBase::activate_integrity_constraints() const {
...
+    _data.x_cut = _data.x_out;
+  } else {
+    for (const auto &[name, value] : _data.x_out) {
+      _data.x_cut[name] = _options.SEPARATION_PARAM * _data.x_out[name] +
+                          (1 - _options.SEPARATION_PARAM) * _data.x_in[name];
+    }
+  }
+}
+
+void BendersBase::compute_invest_cost() {
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
void BendersBase::ComputeInvestCost() {
```

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 10]
```diff
@@ -866,7 +879,7 @@ void BendersBase::ChecksResumeMode() {
     restart_data_printer.Print();
     UpdateMaxNumberIterationResumeMode(last_iter.it);
     benders_timer = Timer(last_iter.benders_elapsed_time);
-    _data.stop = stopping_criterion();
+    _data.stop = should_benders_stop();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/BendersBase.cpp]:
```suggestion
    _data.stop = ShouldBendersStop();
```

File: src/cpp/benders/benders_core/LastIterationReader.cpp [hunk 11]
```diff
@@ -43,7 +43,9 @@ LogData LastIterationReader::GetIterationData(
       last_iteration_file_content_[iteration_name]["subproblem_cost"]
           .asDouble(),
       last_iteration_file_content_[iteration_name]["invest_cost"].asDouble(),
-      x0,
+      {},
+      {},
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/LastIterationReader.cpp:47]: no need to store `x_in` or `x_out`?

File: src/cpp/benders/benders_core/WorkerMaster.cpp [hunk 12]
```diff
@@ -83,34 +83,34 @@ int WorkerMaster::get_number_constraint() const { return _solver->get_nrows(); }
...
+  // cut is -rhs >= alpha  + s^(x-x_cut)
   int ncoeffs(1 + (int)s.size());
   std::vector<char> rowtype(1, 'L');
   std::vector<double> rowrhs(1, 0);
   std::vector<double> matval(ncoeffs, 1);
   std::vector<int> mstart = {0, ncoeffs};
   std::vector<int> mclind(ncoeffs);
 
-  define_rhs_with_master_variable(s, x0, rhs, rowrhs);
+  define_rhs_with_master_variable(s, x_cut, rhs, rowrhs);
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/WorkerMaster.cpp]:
```suggestion
  DefineRhsWithMasterVariable(s, x_cut, rhs, rowrhs);
```

File: src/cpp/benders/benders_core/WorkerMaster.cpp [hunk 13]
```diff
@@ -83,34 +83,34 @@ int WorkerMaster::get_number_constraint() const { return _solver->get_nrows(); }
...
   define_matval_mclind(s, matval, mclind);
 
   solver_addrows(_solver, rowtype, rowrhs, {}, mstart, mclind, matval);
 }
 
 void WorkerMaster::define_rhs_with_master_variable(
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/WorkerMaster.cpp]:
```suggestion
void WorkerMaster::DefineRhsWithMasterVariable(
```

File: src/cpp/benders/benders_core/WorkerMaster.cpp [hunk 14]
```diff
@@ -202,20 +202,21 @@ void WorkerMaster::define_matval_mclind_for_index(
...
+  // cut is -rhs >= alpha  + s^(x-x_cut)
   int ncoeffs(1 + (int)s.size());
   std::vector<char> rowtype(1, 'L');
   std::vector<double> rowrhs(1, 0);
   std::vector<double> matval(ncoeffs, 1);
   std::vector<int> mstart = {0, ncoeffs};
   std::vector<int> mclind(ncoeffs);
 
-  define_rhs_with_master_variable(s, x0, rhs, rowrhs);
+  define_rhs_with_master_variable(s, x_cut, rhs, rowrhs);
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/WorkerMaster.cpp]:
```suggestion
  DefineRhsWithMasterVariable(s, x_cut, rhs, rowrhs);
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 15]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
   bool is_initial_relaxation_requested() const;
   bool switch_to_integer_master(bool is_relaxed) const;
-  void update_trace();
+  virtual void update_trace();
+  void compute_x_cut();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  void ComputeXCut();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 16]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
   virtual void initialize_problems() = 0;
   virtual void init_data();
-  void print_csv();
   void update_best_ub();
-  bool stopping_criterion();
+  bool should_benders_stop();
   bool is_initial_relaxation_requested() const;
   bool switch_to_integer_master(bool is_relaxed) const;
-  void update_trace();
+  virtual void update_trace();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  virtual void UpdateTrace();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 17]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
+  void compute_x_cut();
+  void compute_invest_cost();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  void ComputeInvestCost();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 18]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
+  virtual void compute_ub();
   virtual void get_master_value();
   void getSubproblemCut(SubproblemCutPackage &subproblem_cut_package);
   virtual void post_run_actions() const;
   void build_cut_full(const AllCutPackage &all_package);
-  void deactivate_integrity_constraints() const;
-  void activate_integrity_constraints() const;
-  void set_data_pre_relaxation();
-  void reset_data_post_relaxation();
+  virtual void deactivate_integrity_constraints() const;
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  virtual void DeactivateIntegrityConstraints() const;
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 19]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
+  virtual void activate_integrity_constraints() const;
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  virtual void ActivateIntegrityConstraints() const;
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 20]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
+  virtual void set_data_pre_relaxation();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  virtual void SetDataPreRelaxation();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 21]
```diff
@@ -32,20 +32,22 @@ class BendersBase {
...
+  virtual void reset_data_post_relaxation();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  virtual void ResetDataPostRelaxation();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 22]
```diff
@@ -102,7 +104,8 @@ class BendersBase {
   void print_master_csv(std::ostream &stream, const WorkerMasterDataPtr &trace,
                         Point const &xopt) const;
   void bound_simplex_iter(int simplexiter);
-  bool relaxation_stopping_criterion() const;
+  void update_stopping_criterion();
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  void UpdateStoppingCriterion();
```

File: src/cpp/benders/benders_core/include/BendersBase.h [hunk 23]
```diff
@@ -102,7 +104,8 @@ class BendersBase {
...
+  bool should_relaxation_stop() const;
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/BendersBase.h]:
```suggestion
  bool ShouldRelaxationStop() const;
```

File: src/cpp/benders/benders_core/include/core/ILogger.h [hunk 24]
```diff
@@ -126,6 +129,7 @@ class ILogger {
   virtual void restart_best_iteration(const int best_iterations) = 0;
   virtual void restart_best_iterations_infos(
       const LogData &best_iterations_data) = 0;
+  virtual void log_at_initial_relaxation() = 0;
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/core/ILogger.h]:
```suggestion
  virtual void LogAtInitialRelaxation() = 0;
```

File: src/cpp/benders/benders_core/include/core/ILogger.h [hunk 25]
```diff
@@ -126,6 +129,7 @@ class ILogger {
...
   virtual void log_at_switch_to_integer() = 0;
```
Comment (a-zakir) [inline src/cpp/benders/benders_core/include/core/ILogger.h]:
```suggestion
  virtual void LogAtSwitchToInteger() = 0;
```

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 26]
```diff
@@ -71,6 +71,7 @@ void BendersMpi::step_1_solve_master() {
 void BendersMpi::do_solve_master_create_trace_and_update_cuts() {
   if (_world.rank() == rank_0) {
     if (switch_to_integer_master(_data.is_in_initial_relaxation)) {
+      _logger->log_at_switch_to_integer();
```
Comment (a-zakir) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp]:
```suggestion
      _logger->logAtSwitchToInteger();
```

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 27]
```diff
@@ -92,6 +93,8 @@ void BendersMpi::solve_master_and_create_trace() {
   _logger->display_message("\tSolving master...");
   get_master_value();
   _logger->log_master_solving_duration(get_timer_master());
+
+  compute_x_cut();
```
Comment (a-zakir) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp]:
```suggestion
  ComputeXCut();
```

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 28]
```diff
@@ -220,6 +224,7 @@ void BendersMpi::run() {
     set_cut_storage();
 
     if (is_initial_relaxation_requested()) {
+      _logger->log_at_initial_relaxation();
```
Comment (a-zakir) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp]:
```suggestion
      _logger->LogAtInitialRelaxation();
```

File: src/cpp/benders/benders_sequential/BendersSequential.cpp [hunk 29]
```diff
@@ -81,6 +81,7 @@ void BendersSequential::run() {
   }
 
   if (is_initial_relaxation_requested()) {
+    _logger->log_at_initial_relaxation();
```
Comment (a-zakir) [inline src/cpp/benders/benders_sequential/BendersSequential.cpp]:
```suggestion
    _logger->LogAtInitialRelaxation();
```

File: src/cpp/benders/benders_sequential/BendersSequential.cpp [hunk 30]
```diff
@@ -100,6 +101,7 @@ void BendersSequential::run() {
     get_master_value();
     _logger->log_master_solving_duration(get_timer_master());
 
+    compute_x_cut();
```
Comment (a-zakir) [inline src/cpp/benders/benders_sequential/BendersSequential.cpp]:
```suggestion
    ComputeXCut();
```

File: src/cpp/benders/logger/Master.cpp [hunk 31]
```diff
@@ -83,6 +83,12 @@ void Master::restart_best_iterations_infos(const LogData &best_iteration_data) {
   }
 }
 
+void Master::log_at_initial_relaxation() {
+  for (auto logger : _loggers) {
+    logger->log_at_initial_relaxation();
```
Comment (a-zakir) [inline src/cpp/benders/logger/Master.cpp]:
```suggestion
    logger->LogAtInitialRelaxation();
```

File: src/cpp/benders/logger/Master.cpp [hunk 32]
```diff
@@ -83,6 +83,12 @@ void Master::restart_best_iterations_infos(const LogData &best_iteration_data) {
...
+  }
+}
+
 void Master::log_at_switch_to_integer() {
   for (auto logger : _loggers) {
     logger->log_at_switch_to_integer();
```
Comment (a-zakir) [inline src/cpp/benders/logger/Master.cpp]:
```suggestion
    logger->LogAtSwitchToInteger();
```

File: src/cpp/benders/logger/User.cpp [hunk 33]
```diff
@@ -84,6 +84,11 @@ void User::restart_best_iterations_infos(const LogData &best_iteration_data) {
   log_at_iteration_end(best_iteration_data);
 }
 
+void User::log_at_initial_relaxation() {
```
Comment (a-zakir) [inline src/cpp/benders/logger/User.cpp]:
```suggestion
void User::LogAtInitialRelaxation() {
```

File: src/cpp/benders/logger/User.cpp [hunk 34]
```diff
@@ -84,6 +84,11 @@ void User::restart_best_iterations_infos(const LogData &best_iteration_data) {
...
+  _stream << "--- Switch master formulation to relaxed"
+          << std::endl;
+}
+
 void User::log_at_switch_to_integer() {
```
Comment (a-zakir) [inline src/cpp/benders/logger/User.cpp]:
```suggestion
void User::LogAtSwitchToInteger() {
```

File: src/cpp/benders/logger/UserFile.cpp [hunk 35]
```diff
@@ -105,6 +105,13 @@ void UserFile::restart_best_iterations_infos(
   log_at_iteration_end(best_iteration_data);
 }
 
+void UserFile::log_at_initial_relaxation() {
```
Comment (a-zakir) [inline src/cpp/benders/logger/UserFile.cpp]:
```suggestion
void UserFile::LogAtInitialRelaxation() {
```

File: src/cpp/benders/logger/UserFile.cpp [hunk 36]
```diff
@@ -105,6 +105,13 @@ void UserFile::restart_best_iterations_infos(
...
+  _file << LINE_PREFIX
+        << "--- Switch master formulation to relaxed"
+        << std::endl;
+  _file.flush();
+}
+
 void UserFile::log_at_switch_to_integer() {
```
Comment (a-zakir) [inline src/cpp/benders/logger/UserFile.cpp]:
```suggestion
void UserFile::LogAtSwitchToInteger() {
```

File: src/cpp/benders/logger/include/logger/Master.h [hunk 37]
```diff
@@ -49,6 +49,7 @@ class Master : public ILogger {
   void restart_best_iteration(const int best_iterations) override;
   void restart_best_iterations_infos(
       const LogData &best_iterations_data) override;
+  void log_at_initial_relaxation() override;
   void log_at_switch_to_integer() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/Master.h]:
```suggestion
  void LogAtSwitchToInteger() override;
```

File: src/cpp/benders/logger/include/logger/Master.h [hunk 38]
```diff
@@ -49,6 +49,7 @@ class Master : public ILogger {
   void restart_best_iteration(const int best_iterations) override;
   void restart_best_iterations_infos(
       const LogData &best_iterations_data) override;
+  void log_at_initial_relaxation() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/Master.h]:
```suggestion
  void LogAtInitialRelaxation() override;
```

File: src/cpp/benders/logger/include/logger/User.h [hunk 39]
```diff
@@ -41,6 +41,7 @@ class User : public ILogger {
   void restart_best_iteration(const int best_iterations) override;
   void restart_best_iterations_infos(
       const LogData &best_iterations_data) override;
+  void log_at_initial_relaxation() override;
   void log_at_switch_to_integer() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/User.h]:
```suggestion
  void LogAtSwitchToInteger() override;
```

File: src/cpp/benders/logger/include/logger/User.h [hunk 40]
```diff
@@ -41,6 +41,7 @@ class User : public ILogger {
   void restart_best_iteration(const int best_iterations) override;
   void restart_best_iterations_infos(
       const LogData &best_iterations_data) override;
+  void log_at_initial_relaxation() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/User.h]:
```suggestion
  void LogAtInitialRelaxation() override;
```

File: src/cpp/benders/logger/include/logger/UserFile.h [hunk 41]
```diff
@@ -40,7 +40,7 @@ class UserFile : public ILogger {
   void restart_best_iteration(const int best_iterations) override;
   void restart_best_iterations_infos(
       const LogData &best_iterations_data) override;
-  
+  void log_at_initial_relaxation() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/UserFile.h]:
```suggestion
  void LogAtInitialRelaxation() override;
```

File: src/cpp/benders/logger/include/logger/UserFile.h [hunk 42]
```diff
@@ -40,7 +40,7 @@ class UserFile : public ILogger {
...
   void log_at_switch_to_integer() override;
```
Comment (a-zakir) [inline src/cpp/benders/logger/include/logger/UserFile.h]:
```suggestion
  void LogAtSwitchToInteger() override;
```

File: tests/cpp/TestDoubles/LoggerStub.h [hunk 43]
```diff
@@ -24,5 +24,6 @@ class LoggerNOOPStub: public ILogger {
   void restart_best_iteration(const int best_iterations) override {}
   void restart_best_iterations_infos(
       const LogData& best_iterations_data) override {}
+  void log_at_initial_relaxation() override {}
```
Comment (a-zakir) [inline tests/cpp/TestDoubles/LoggerStub.h]:
```suggestion
  void LogAtInitialRelaxation() override {}
```

File: tests/cpp/TestDoubles/LoggerStub.h [hunk 44]
```diff
@@ -24,5 +24,6 @@ class LoggerNOOPStub: public ILogger {
...
   void log_at_switch_to_integer() override {}
```
Comment (a-zakir) [inline tests/cpp/TestDoubles/LoggerStub.h]:
```suggestion
  void LogAtSwitchToInteger() override {}
```

**Review by a-zakir — CHANGES_REQUESTED** [review]

File: src/cpp/benders/benders_core/LastIterationReader.cpp (hunk 11 above)
Comment (tbittar) [reply src/cpp/benders/benders_core/LastIterationReader.cpp:47]: Indeed no need to store `x_in` or `x_out` as only `x_cut` is useful for the user, and writing `x_in` and `x_out` in the last iteration file would be heavy (this is the list of candidates with an invested value) and confusing

**Review by JasonMarechal25 — APPROVED** [review]

**Review by a-zakir — APPROVED** [review]

**Pull request merged**